torch/testing/_internal/mps_utils.py

```diff
@@ -0,0 +1,988 @@
+    ),
+    "signal.windows.cosine": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.exponential": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.gaussian": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.general_cosine": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.general_hamming": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.hamming": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.hann": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]),
+    "signal.windows.kaiser": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "signal.windows.nuttall": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float32]
+    ),
+    "eye": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16, torch.float32]),
+    # trunc_tensor not working properly for float16
+    "divtrunc_rounding": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]),
+    "fmod": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]),
+    # atomic operation in backward pass
+    "_unsafe_masked_index": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]),
+    "_unsafe_masked_index_put_accumulate": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]
+    ),
+    # Same issue as `argsort` and `sort` with duplicate elements (undefined behaviour).
+    # Forward pass is passing since `msort` doesn't return the indices, just the values, which match the CPU.
+    # On the backward pass for `sort` both are used (values and indices), thus resulting in a issmatch between CPU and MPS.
+    # Running `msort` with stable `sort` passes.
+    "msort": MPSSkipInfo(TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]),
+    "nn.functional.pairwise_distance": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16]
+    ),
+    "nn.functional.conv1d": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16], lower=14.0
+    ),
+    "nn.functional.conv2d": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16], lower=14.0
+    ),
+    "nn.functional.conv3d": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH, dtypes=[torch.float16, torch.float32], lower=14.0
+    ),
+    # Uncategorized grad failures
+    "nanquantile": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH,
+        dtypes=[torch.float32],
+        upper=15.0,
+    ),
+    "quantile": MPSSkipInfo(
+        TEST_OUTPUT_GRAD_MATCH,
+        dtypes=[torch.float32],
+        upper=15.0,
+    ),
+}
+
+COMPLEX_DTYPES = [torch.complex32, torch.complex64]
+"""Ops which do not have support for complex dtypes and are expected to fail"""
+"""Only applies to MacOS14+, all complex ops are unsupported on earlier MacOS versions"""
```

Comment:
This is not a valid statement, see https://developer.apple.com/documentation/metalperformanceshaders/mpsdatatype/mpsdatatypecomplexfloat32?language=objc, complex was introduced in MacOS-13